### PR TITLE
Fix issue 781

### PR DESCRIPTION
**Pull request by RAVEENSR**
## Purpose
This PR fixes https://github.com/wso2/analytics-apim/issues/781 issue.

## Security checks
 - Followed secure coding standards in http://wso2.com/technical-reports/wso2-secure-engineering-guidelines? yes
 - Ran FindSecurityBugs plugin and verified report? yes
 - Confirmed that this PR doesn't commit any keys, passwords, tokens, usernames, or other secrets? yes

### CODE REVIEW

**Comment by RAVEENSR** [comment]
@fazlan-nazeem Please review.

File: components/org.wso2.analytics.apim.idp.client/src/main/java/org/wso2/analytics/apim/idp/client/ApimIdPClient.java [hunk 1]
```diff
@@ -403,6 +392,12 @@ private boolean isOAuthApplicationExists(String oAuthAppName) throws IdPClientEx
                 if (IdPClientConstants.PASSWORD_GRANT_TYPE.equals(grantType)) {
                     tokenCache.put(oAuth2TokenInfo.getAccessToken(),
                             new ExternalSession(username, oAuth2TokenInfo.getAccessToken()));
+                    TokenData tokenData = new TokenData(
```
Comment (ruks) [inline components/org.wso2.analytics.apim.idp.client/src/main/java/org/wso2/analytics/apim/idp/client/ApimIdPClient.java:395]: did we format this correctly?

Comment (RAVEENSR) [reply components/org.wso2.analytics.apim.idp.client/src/main/java/org/wso2/analytics/apim/idp/client/ApimIdPClient.java:395]: Yes, we cannot fit all the parameters of the `TokenData` constructor in the single line. Hence, in order to improve the readability parameters are given one per line one below another.

File: components/org.wso2.analytics.apim.idp.client/src/main/java/org/wso2/analytics/apim/idp/client/ApimIdPClient.java [hunk 2]
```diff
@@ -435,6 +430,15 @@ private boolean isOAuthApplicationExists(String oAuthAppName) throws IdPClientEx
         if (!this.oAuthAppInfoMap.containsKey(oAuthAppContext)) {
             oAuthAppContext = ApimIdPClientConstants.DEFAULT_SP_APP_CONTEXT;
         }
+        ExternalSession session = tokenCache.getIfPresent(token);
+        String username;
+        if (session == null) {
+            LOG.info("Token is not available in the Token Cache. Hence, username cannot be retrieved. " +
```
Comment (praminda) [inline components/org.wso2.analytics.apim.idp.client/src/main/java/org/wso2/analytics/apim/idp/client/ApimIdPClient.java:436]: Do we need this descriptive info log? How above simply saying somthing like below
`Token in the token data map will be deleted as the token is not available in the Token Cache`

File: components/org.wso2.analytics.apim.idp.client/src/main/java/org/wso2/analytics/apim/idp/client/token/TokenDataHolder.java [hunk 3]
```diff
@@ -0,0 +1,77 @@
...
+    /**
+     * This method can be used to get the singleton instance of this class.
+     * @return the singleton instance.
+     */
+    public static TokenDataHolder getInstance() {
+        return instance;
+    }
+
+    /**
+     *  Adds a given key, value pair to the map.
```
Comment (praminda) [inline components/org.wso2.analytics.apim.idp.client/src/main/java/org/wso2/analytics/apim/idp/client/token/TokenDataHolder.java:48]:
```diff
-- *  Adds a given key, value pair to the map.
++ * Adds a given key, value pair to the map.
```

File: components/org.wso2.analytics.apim.idp.client/src/main/java/org/wso2/analytics/apim/idp/client/token/TokenDataHolder.java [hunk 4]
```diff
@@ -0,0 +1,77 @@
...
+     */
+    public void addTokenDataToMap(String key, TokenData value) {
+        if (key != null && value != null) {
+            tokenDataMap.put(key, value);
+            LOG.debug("Adding token data key, value pair to the token map :" + key + " , " + value.toString());
+        }
+    }
+
+    /**
+     *  Removes a given key, value pair from the map.
```
Comment (praminda) [inline components/org.wso2.analytics.apim.idp.client/src/main/java/org/wso2/analytics/apim/idp/client/token/TokenDataHolder.java:60]:
```diff
-- *  Removes a given key, value pair from the map.
++ * Removes a given key, value pair from the map.
```

File: components/org.wso2.analytics.apim.idp.client/src/main/java/org/wso2/analytics/apim/idp/client/token/TokenDataMapCleaner.java [hunk 5]
```diff
@@ -0,0 +1,70 @@
...
+import org.apache.commons.logging.Log;
+import org.apache.commons.logging.LogFactory;
+
+import java.util.Iterator;
+import java.util.Map;
+import java.util.Timer;
+import java.util.TimerTask;
+
+/**
+ *  Responsible to remove expired access tokens from the tokenData map.
```
Comment (praminda) [inline components/org.wso2.analytics.apim.idp.client/src/main/java/org/wso2/analytics/apim/idp/client/token/TokenDataMapCleaner.java:29]:
```diff
-- *  Responsible to remove expired access tokens from the tokenData map.
++ * Responsible to remove expired access tokens from the tokenData map.
```

File: components/org.wso2.analytics.apim.idp.client/src/main/java/org/wso2/analytics/apim/idp/client/ApimIdPClient.java (hunk 2 above)
Comment (RAVEENSR) [reply components/org.wso2.analytics.apim.idp.client/src/main/java/org/wso2/analytics/apim/idp/client/ApimIdPClient.java:436]: This code segment is no longer there. This is now replaced with another code segment.

File: components/org.wso2.analytics.apim.idp.client/src/main/java/org/wso2/analytics/apim/idp/client/token/TokenDataMapCleaner.java (hunk 5 above)
Comment (RAVEENSR) [reply components/org.wso2.analytics.apim.idp.client/src/main/java/org/wso2/analytics/apim/idp/client/token/TokenDataMapCleaner.java:29]: Fixed in f2f5dac89813ed61d4af68d5da8ca54e160d14a4

File: components/org.wso2.analytics.apim.idp.client/src/main/java/org/wso2/analytics/apim/idp/client/token/TokenDataHolder.java (hunk 4 above)
Comment (RAVEENSR) [reply components/org.wso2.analytics.apim.idp.client/src/main/java/org/wso2/analytics/apim/idp/client/token/TokenDataHolder.java:60]: Fixed in f2f5dac89813ed61d4af68d5da8ca54e160d14a4

File: components/org.wso2.analytics.apim.idp.client/src/main/java/org/wso2/analytics/apim/idp/client/token/TokenDataHolder.java (hunk 3 above)
Comment (RAVEENSR) [reply components/org.wso2.analytics.apim.idp.client/src/main/java/org/wso2/analytics/apim/idp/client/token/TokenDataHolder.java:48]: Fixed in f2f5dac89813ed61d4af68d5da8ca54e160d14a4

**Pull request merged**